net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,314 @@
+ *
+ *
+ */
+
+
+struct gkprioq_sched_data {
+	    /* Parameters */
+	u32 max_limit;
+
+            /* Queue state */
+	struct Qdisc *prio_array[MAX_PRIORITY];
+	u16 highest_prio;
+	u16 lowest_prio;
+};
+
+static inline u16 calc_new_high_prio(const struct gkprioq_sched_data *q)
+{
+	int prio;
+
+	for (prio = q->highest_prio - 1; prio >= q->lowest_prio; prio--) {
+		if ((q->prio_array[prio])->q.qlen > 0)
+			return prio;
+	}
+
+	return 0; /* GK queue is empty, return 0 
+		   * (default highest priority setting). */
+}
+
+static inline u16 calc_new_low_prio(const struct gkprioq_sched_data *q)
+{
+	int prio;
+
+	if (q->lowest_prio == 0)
+		return 0;
+
+	for (prio = q->lowest_prio + 1; prio <= q->highest_prio; prio++) {
+		if ((q->prio_array[prio])->q.qlen > 0)
+			return prio;
+	}
+	
+	return 0; /* GK queue is empty, return 0 
+		   * (default lowest priority setting). */
+}
+
+static int gkprioq_enqueue(struct sk_buff *skb, struct Qdisc *sch,
+			   struct sk_buff **to_free)
+{
+	struct Qdisc *your_queue;
+	int ret, wlen, gk_len;
+	u16 your_prio, lp; 
+
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	wlen = skb_network_offset(skb);
+
+	switch(tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = (ipv4_get_dsfield(ip_hdr(skb))) >> 2;
+		break;
+
+	case htons(ETH_P_IPV6):
+	 	wlen += sizeof(struct ipv6hdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = (ipv6_get_dsfield(ipv6_hdr(skb))) >> 2;
+	      	break;
+
+	default: 
+		your_prio = 0;
+		break;
+	}
+
+	/* Check to update highest and lowest priorities. */
+
+	if (your_prio > q->highest_prio)
+		q->highest_prio = your_prio;
+
+	if (your_prio < q->lowest_prio)
+		q->lowest_prio = your_prio;	
+			
+	your_queue = q->prio_array[your_prio];
+	lp = q->lowest_prio;
+
+	gk_len = sch->q.qlen;
+	if (gk_len < q->max_limit) {
+
+		ret = qdisc_enqueue(skb, your_queue, to_free);
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+			return ret;
+		}
+
+		qdisc_qstats_backlog_inc(sch, skb);
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+	}
+
+	/* Look to lowest priority queue to check if ingress packet can be 
+	 * accomodated. Otherwise no choice but to drop ingress packet.
+	 */
+ 
+	if (lp != your_prio) { 		
+		
+		__qdisc_queue_drop_head(q->prio_array[lp], 
+				      &(q->prio_array[lp])->q, to_free);
+		qdisc_qstats_drop(sch);
+		
+		/* update lowest priority field. */
+
+		if ((q->prio_array[lp])->q.qlen == 0)
+			q->lowest_prio = calc_new_low_prio(q); 
+
+		ret = qdisc_enqueue(skb, your_queue, to_free);
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+				return ret;
+		}
+
+		qdisc_qstats_backlog_inc(sch, skb);
+		return NET_XMIT_SUCCESS;	
+	}
+	
+	return qdisc_drop(skb, sch, to_free);
+	
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
```

Comment:
`__NET_XMIT_BYPASS` is saying that the packet should try to call this function again because it could result in a different result at that time. Yes, it should work.